Adding extra fitting methods for quantum environments

gsuarezr opened the pull request:
**Description**
This PR aims to add other methods used to generate decaying exponential decompositions of environments. The methods added in this PR have been applied to fermionic environments in the literature, while the current draft PR only implements these methods for  bosonic environments, hopefully they can be used in fermionic environments soon. The key issue there is how to create the interface for the creation of arbitrary fermionic environments. 

The methods this PR aims to add are : 

- [x] AAA 
- [x] Prony
- [x] Matrix Pencil
- [x] Esprit
- [x] Espira (I and II)
- [x]  Power Spectrum NLSQ

**Checklist**

- [x] Refactor fitting methods
- [x] Add unit tests for fitting methods
- [x] Include the changelog 
- [x] Document characteristics and when to use each method


NOTE: I decided to not do the balanced truncation method in this PR, it works to reduce exponents in all of them, but I probably need some time to understand it

**Related issues or PRs** 
#2580 

I would like somebody else's opinion on whether or not to keep mp, and espira-I, while they have their uses in signal analysis, maybe not so much here.  In  qutip/qutip-tutorials#72  1d all of them are used. It needs work as a tutorial, but maybe someone can run it and emit their opinion

hodgestar commented on qutip/core/environment.py:
```diff
@@ -767,6 +768,133 @@ def approx_by_sd_fit(
             ckAR, vkAR, ckAI, vkAI, combine=combine, T=self.T, tag=tag)
         return approx_env, fit_info
 
+    def approx_by_aaa(
```
It looks like we might have an every growing number of such methods so perhaps we should have one method named `approximate` that accepts a method name and dispatches the rest of the arguments to the implementing functions. For example, in a simple implementation `.approximate("aaa", ...)` might end up calling `_approx_by_aaa(...)`.

gsuarezr replied:
Hi Simon, 

I wasn´t expecting any comment so soon, thanks a lot!

Indeed, there seems to be a growing number of methods appearing in the literature. I added the ones in [this review article](https://pubs.aip.org/aip/jcp/article-abstract/160/20/204105/3294390/High-accuracy-exponential-decomposition-of-bath?redirectedFrom=fulltext) (sorry to attach a paper behind a paywall sadly no arxiv 😢 ). If @pmenczel  agrees then I think that's probably the best solution, I guess the main issue would be that some methods don´t take in the same parameters so it might be confusing for a user but I think that's fine 😄 

To be honest for my purposes only Prony and AAA are of importance, but since the other ones are similar I included them as well, but not 100% sure all should be included in the end.

pmenczel replied:
In principle I would like an implementation like `.approximate("aaa", ...)`. That is actually how I implemented it at first. However, I found it very annoying to use since all the methods require different parameters, and the IDE then didn't show a tooltip listing the required parameters. That is why I went with having different functions.

Perhaps that could be fixed using `@overload`, though?

gsuarezr replied:
I played around with overload in the last week, and I could get the hinting working for simple cases, but when doing it similarly in the code we already have for some reason it no longer works, at least not in vscode, I'm still looking into it though! 
Sorry for not replying before 😢

pmenczel commented:
Hi, thank you for all your work. I just pushed a commit without major changes to code or text, but moving a lot of documentation around. Take a look to see if you like it. The problem was that `@overload` versions of functions are not supposed to have docstrings, and these docstrings are ignored in the automatically generated API documentation. So I created a central place in the docs to document all of the approximation methods.

In addition, I thought it would make sense to have the matsubara and pade approximations also available through `approximate`. Unfortunately, if a subclass adds new @overload versions of a functions, all previous ones must also be copied into the subclass. Codeclimate is a bit unhappy about that, but it seems unavoidable with the single-approximate-function approach.

A few other points:
* I left two TODOs in the users guide
* I am unsure about `mp` and `espira-I`. If even you don't know a use case, we could consider dropping it to reduce the amount of code and not overwhelm the user with choices. On the other hand, you have already written the code, so it would seem a bit stupid to delete it now. I would say, it is up to you.
* Regarding the AAA, we could add a little comment in the code that our own implementation can perhaps be removed once we support only scipy 1.15+.
* I am unsure about that $r^2$ coefficient instead of least squares. Are you saying that it improves the results significantly?
    - Perhaps we could have a flag to decide whether to do $r^2$ or least squares?
    - There are many places where we refer to least squares, like docstrings, name of `rmse` variable, referring to fitting functions as "NLSQ", ...
    - Double check that the description of the `fit_info` output is correct everywhere in the new section in the Users Guide. By the way, it seems to be missing in the `approximate("aaa", ...)` docstring.
* Let's also double check if stuff has to be updated in the HEOM guide.
* I haven't touched any fermions stuff, I'm leaving that for the follow-up fermion fitting PR that you mentioned.

Note to myself: I still have to have a closer look at the tutorial notebooks

gsuarezr commented:
Thanks ! and thak you for such a quick review! Indeed it's a shame that codeclimate doesn't like it. probably going to be repeated when fermionic things are added, but I guess we can ignore codeclimate in this particular case?

> I left two TODOs in the users guide

Cool I'll take a look at them 

> I am unsure about mp and espira-I. If even you don't know a use case, we could consider dropping it to reduce the amount of code and not overwhelm the user with choices. On the other hand, you have already written the code, so it would seem a bit stupid to delete it now. I would say, it is up to you.

Yeah, I think more than the amount of code is too many choices for the user, right now I think we can just delete mp, and keep espira-I because it seems to be working nicely with fermions, though I need to take a look at some more examples, of course as I mentioned on line, whenever pade is available pade is better, a lot of the time by a lot

> Regarding the AAA, we could add a little comment in the code that our own implementation can perhaps be removed once we support only scipy 1.15+.

Yes, we definitely should. The main problem is that the current implementation of scipy does not return what I called iindices, so we cannot construct the objects needed for espira from it. But I guess we can ask for it! perhaps They'll add esprit and those prony based methods in the future, I asked about it on the scientific python discord, I'm waiting for a reply


> I am unsure about that $1-r^{2}$ coefficient instead of least squares. Are you saying that it improves the results significantly?

it's supposed to replace rmse,  both are measures of the goodness of fit. That one just can be kept as it is, $r^{2}$ just seems to be a more common measure so I wanted to replace rmse, but probably did not document that correctly, it doesn't change the results at all.

On that hand I was thinking maybe, the goodness of fit is not so informative sometimes, because you might have a good fit for the correlation function, but that does not reproduce the power spectrum so well (The bit about problems keeping kms with these approximations, I think that was briefly discussed in Neill's pseudomode paper). So perhaps goodness of fit to both should be reported? Not sure. Maybe it's just better to keep it as it was.

I will just delete the $r^{2}$ stuff


> Double check that the description of the fit_info output is correct everywhere in the new section in the Users Guide. By the way, it seems to be missing in the approximate("aaa", ...) docstring. Let's also double check if stuff has to be updated in the HEOM guide.

Sure I'll check the guides and dosctrings, I didn't check them too much in case many changes were required. I think the notation in the summaries is also not uniform right now, I'll just keep an eye on it to make it uniform and check stuff in the user's guide

> I haven't touched any fermions stuff, I'm leaving that for the follow-up fermion fitting PR that you mentioned.

Sure, I'll just create that one once this one is merged.


> Note to myself: I still have to have a closer look at the tutorial notebooks

If notation, is going to be kept (up to minor details, like I just noticed sometimes is Nmax or N_max in different methods), then I will take another look at the tutorials and send you a message on line once they're ready

pmenczel commented:
> Yeah, I think more than the amount of code is too many choices for the user, right now I think we can just delete mp, and keep espira-I because it seems to be working nicely with fermions, though I need to take a look at some more examples, of course as I mentioned on line, whenever pade is available pade is better, a lot of the time by a lot

Sounds reasonable :)

> Yes, we definitely should. The main problem is that the current implementation of scipy does not return what I called iindices, so we cannot construct the objects needed for espira from it. But I guess we can ask for it! perhaps They'll add esprit and those prony based methods in the future, I asked about it on the scientific python discord, I'm waiting for a reply

Oh in that case, let's definitely keep your implementation for now, and if the scipy implementation can't do everything we need, there's no need to add a comment. But if they reply and eventually change things, that would be nice!

> it's supposed to replace rmse, both are measures of the goodness of fit. That one just can be kept as it is, r2 just seems to be a more common measure so I wanted to replace rmse, but probably did not document that correctly, it doesn't change the results at all.
> 
> On that hand I was thinking maybe, the goodness of fit is not so informative sometimes, because you might have a good fit for the correlation function, but that does not reproduce the power spectrum so well (The bit about problems keeping kms with these approximations, I think that was briefly discussed in Neill's pseudomode paper). So perhaps goodness of fit to both should be reported? Not sure. Maybe it's just better to keep it as it was.
> 
> I will just delete the r2 stuff

Ah. Well if r2 is a more common thing to do in fitting, it's okay to keep it. I just didn't know, I thought least squares was the "normal" thing. Just make sure that everything is named and documented right.

I think if we fit the power spectrum, we only need to include information about the goodness of how well it fits the power spectrum; if we fit the CF we only need to include the goodness of how it fits the CF, etc.

gsuarezr commented:
In the end I removed $r^{2}$, it is more common but having one of the two is enough since [they both agree on what the better model is](https://datascience.stackexchange.com/questions/100605/rmse-vs-r-squared ). I also removed mp, and did the todos in the user guide. I left a comment there in case what I wrote wasn't clear

> I think if we fit the power spectrum, we only need to include information about the goodness of how well it fits the power spectrum; if we fit the CF we only need to include the goodness of how it fits the CF, etc.

Sure, that's sort of the logical thing, I was just thinking about it because I've found that these goodness of fit measures are not too useful to know which fit is better in many cases because of the KMS stuff

pmenczel commented on qutip/utilities.py:
```diff
@@ -523,3 +529,414 @@ def _fit(fun, num_params, xdata, ydata, guesses, lower, upper, sigma,
+    return amp * np.power(phase, np.arange(len(orig)))
+
+
+def prony_methods(method: str, signal: ArrayLike, n: int):
```
```
method: Literal["prony", "esprit"]
```

Also, the formatting of the docstring below is off. Same for espira.

pmenczel commented on qutip/utilities.py:
```diff
@@ -523,3 +529,414 @@ def _fit(fun, num_params, xdata, ydata, guesses, lower, upper, sigma,
+        method (str): The method to obtain the roots of the prony polynomial
+        can be prony, and Estimation of signal parameters
+        via rotational invariant techniques (esprit)
+        signal (np.ndarray): The input signal (1D complex array).
+        n (int): Desired number of modes to  use as estimation
```
I'm not so familiar with this fitting stuff. For prony and espira, the sampling points are always assumed to be evenly spaced?

gsuarezr replied:
Yes!

pmenczel commented on qutip/tests/test_utilities.py:
```diff
@@ -127,19 +127,43 @@ class TestFitting:
         return np.real(a * np.exp(-(b + 1j * c) * x))
 
-    def test_same_model(self):
+    @pytest.fixture
+    def generate_data(self):
```
Now that the fitting is moving a bit more in the spotlight, it might be good to test a little better. How about something like

```suggestion
    @pytest.fixture
    @pytest.mark.parametrize("noisy", [False, True])
    def generate_data(self, noisy):
```

and add some small random noise to `y`?

gsuarezr replied:
Sounds Good!

pmenczel commented on qutip/tests/test_utilities.py:
```diff
@@ -127,19 +127,43 @@ class TestFitting:
+        x, y, _, _ = generate_data
+        result = utils.aaa(y, x, tol=1e-13, max_iter=4)
+        rmse = result["rmse"]
+        assert rmse < 1e-3
```
For all the new fitting methods: in addition to checking that the returned rmse is small enough, it would be good to also check that the result is actually close to the original data (`np.assert_allclose`)

gsuarezr replied:
Sure, will check for that as well

pmenczel commented on qutip/tests/core/test_environment.py:
```diff
@@ -624,6 +625,71 @@ def test_dynamic_sd_fit(self, reference, wMax, tol, params):
+        pytest.param(UDReference(1, .5, .1, 1), 2, 1e-3, id='UD Example'),
+    ])
+    def test_fixed_prony_fit(self, reference, tMax, tol):
+        # TODO: come up with a better test, for now I was just fitting into this
+        # style but those support points are not good for a low number of
+        # exponents
```
TODO here. Also, tests for ESPIRA fits are missing, right?

gsuarezr replied:
yes, the test need to be better I just added more exponents because I didn't want to test those examples at the time. I was planning to do it later on in case names or arguments changed. I will delete the TODO and add the ones for ESPIRA

pmenczel commented on doc/guide/environments/approximations.rst:
```diff
@@ -0,0 +1,666 @@
+     - Partially
+     - Yes
+
+In the table requires extra information refers to whether we need to input 
```
```suggestion
In the table, "Requires Extra Information" refers to whether we need to input 
```
(etc)

pmenczel reviewed (commented):
Thank you for your changes! I think we are getting very close. I have some comments with suggestions on naming and some docstrings. Other than that, only:

* There are still some places in docstrings etc that refer to the "approx_by_" functions, we should update to "approximate"
* Same for the HEOM users guide maybe?
* Out of curiosity, did you ever hear back from the scipy discord about AAA?
* Again, note to myself to look at the notebook 👀

gsuarezr commented:
> Out of curiosity, did you ever hear back from the scipy discord about AAA?

No, they said they'd look into it 😢

gsuarezr commented on qutip/utilities.py:
```diff
@@ -523,3 +529,414 @@ def _fit(fun, num_params, xdata, ydata, guesses, lower, upper, sigma,
+    params = _unpack(
+        np.array([val for pair in zip(amplitudes, phases) for val in pair]), 2)
+
+    rmse = _rmse(_prony_model, signal, signal, params)
```
I wasn't sure which bit looked fishy

The first two lines it's just because I have  two lists $[c_{1},c_{2},....]$ and $[p_{1},p_{2},....]$ so the [val for pair in zip(amplitudes, phases) for val in pair] is to have $[c_{1},p_{1},c_{2},p_{2}....]$ and the unpack to have $[[c_{1},p_{1}],[c_{2},p_{2}]....]$. Mostly so params has the same structure as the least squares stuff. 

if it's the fact that signal appears twice, it's just because the length of the original signal is needed to compute the approximated one, because the date is evenly spaced. I could have passed len(signal) as xdata and modify the function instead from 
```python
 def _prony_model(orig, amp, phase):
      return amp * np.power(phase, np.arange(len(orig)))
```
to
```python
 def _prony_model(n, amp, phase):
      return amp * np.power(phase, np.arange(n))
```

I guess that way is more clear

pmenczel replied:
Ah sorry, yes I was referring to the double "signal". I only meant to quote that line, but I think github always includes at least 4 lines. Thank you for the change!

gsuarezr commented on qutip/utilities.py:
```diff
@@ -361,7 +363,7 @@ def iterated_fit(
     sigma: float | ArrayLike = None,
     maxfev: int = None
-) -> tuple[float, ArrayLike]:
+) -> tuple[float, float, ArrayLike]:
```
True, forgot to delete it when I removed $r^{2}$

gsuarezr commented on qutip/utilities.py:
```diff
@@ -523,3 +529,414 @@ def _fit(fun, num_params, xdata, ydata, guesses, lower, upper, sigma,
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
+def aaa(func: Callable[..., complex], z: ArrayLike,
+        tol: float = 1e-13, max_iter: int = 100):
```
Could we avoid adding the _fit?  I sort of don't like it because except for AAA all I did in utilities is phase estimation I get the exponent later in the environment file.  I also think when you add fit people immediately think minimization and that's not it here

gsuarezr commented:
I changed the Ohmic environment reference because the value we were using at zero temperature was only ok for Ohmic, but not super or sub Ohmic. The problem was we were using 

$C(t)=\alpha \omega_{c}^{s+1} \Gamma(s+1) (1+ i \omega_{c} t)^{-(s+1)}$

But the right one is 

$C(t)=\alpha \omega_{c}^{2} \Gamma(s+1) (1+ i \omega_{c} t)^{-(s+1)}$

so it was only good for $s=1$, Ohmic, whenever $s \neq 1 $. The correlation function for $T=0$ and $T=1\times 10^{-30}$ would differ significantly, now that's fixed :)

The other unrelated thing I did was add a method called rescale. The method multiplies the coefficients of the environment times a constant. I think this is useful when running multiple simulations that only differ by the coupling constant, so that fits do not need to be performed several times, but if it feels weird, the implementation is not good, or it seems confusing feel free to delete it
